AzureStackHub: panic: runtime error: slice bounds out of range

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->

Hey there, I'm currently working on getting OpenShift 4.9 running on Azure StackHub. We already have a test cluster running and I'm currently looking into running the [azure-disk-csi-driver-controller](https://github.com/kubernetes-sigs/azuredisk-csi-driver) and making sure all e2e tests in the [CSI suite](https://github.com/kubernetes/kubernetes/blob/588d98440706e317ca0fcc5b510285caaf161e51/test/e2e/storage/framework/testpattern.go) pass.

**What happened**:

On some tests (especially the multivolume ones), the driver dies with the following panic:

```go
I0629 14:07:35.520114       1 utils.go:96] GRPC request: {"volume_id":"/subscriptions/e90a4bc2-2f4e-4d36-874b-9011e4734d74/resourceGroups/padillon-7xkrc-rg/providers/Microsoft.Compute/disks/pvc-1ecb4793-f484-4914-821d-bbee3164fed9"}
I0629 14:07:35.520201       1 controllerserver.go:349] deleting azure disk(/subscriptions/e90a4bc2-2f4e-4d36-874b-9011e4734d74/resourceGroups/padillon-7xkrc-rg/providers/Microsoft.Compute/disks/pvc-1ecb4793-f484-4914-821d-bbee3164fed9)
I0629 14:07:35.535709       1 utils.go:95] GRPC call: /csi.v1.Controller/DeleteVolume
I0629 14:07:35.535727       1 utils.go:96] GRPC request: {"volume_id":"/subscriptions/e90a4bc2-2f4e-4d36-874b-9011e4734d74/resourceGroups/padillon-7xkrc-rg/providers/Microsoft.Compute/disks/pvc-8c4595fd-d372-4d45-9026-1e2b4fc5b7bc"}
I0629 14:07:35.535812       1 controllerserver.go:349] deleting azure disk(/subscriptions/e90a4bc2-2f4e-4d36-874b-9011e4734d74/resourceGroups/padillon-7xkrc-rg/providers/Microsoft.Compute/disks/pvc-8c4595fd-d372-4d45-9026-1e2b4fc5b7bc)
I0629 14:07:35.577065       1 controllerserver.go:351] delete azure disk(/subscriptions/e90a4bc2-2f4e-4d36-874b-9011e4734d74/resourceGroups/padillon-7xkrc-rg/providers/Microsoft.Compute/disks/pvc-1ecb4793-f484-4914-821d-bbee3164fed9) returned with Retriable: false, RetryAfter: 0s, HTTPStatusCode: 404, RawError: Retriable: false, RetryAfter: 0s, HTTPStatusCode: 404, RawError: {"error":{"code":"ResourceNotFound","message":"The Resource 'Microsoft.Compute/disks/pvc-1ecb4793-f484-4914-821d-bbee3164fed9' under resource group 'padillon-7xkrc-rg' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix"}}
E0629 14:07:35.577092       1 utils.go:100] GRPC error: Retriable: false, RetryAfter: 0s, HTTPStatusCode: 404, RawError: Retriable: false, RetryAfter: 0s, HTTPStatusCode: 404, RawError: {"error":{"code":"ResourceNotFound","message":"The Resource 'Microsoft.Compute/disks/pvc-1ecb4793-f484-4914-821d-bbee3164fed9' under resource group 'padillon-7xkrc-rg' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix"}}
I0629 14:07:35.590553       1 azure_controller_standard.go:159] azureDisk - detach disk: name "pvc-b3e2278e-e8a7-450f-beed-bd65a40634e2" uri "/subscriptions/e90a4bc2-2f4e-4d36-874b-9011e4734d74/resourcegroups/padillon-7xkrc-rg/providers/microsoft.compute/disks/pvc-b3e2278e-e8a7-450f-beed-bd65a40634e2"
I0629 14:07:35.590581       1 azure_controller_standard.go:159] azureDisk - detach disk: name "pvc-6c570f18-08df-46a0-a471-6c8bb12dcbd0" uri "/subscriptions/e90a4bc2-2f4e-4d36-874b-9011e4734d74/resourcegroups/padillon-7xkrc-rg/providers/microsoft.compute/disks/pvc-6c570f18-08df-46a0-a471-6c8bb12dcbd0"
panic: runtime error: slice bounds out of range [2:1]

goroutine 126 [running]:
sigs.k8s.io/cloud-provider-azure/pkg/provider.(*availabilitySet).DetachDisk(0xc000010da0, 0xc00060c570, 0x17, 0xc0003a7530, 0x0, 0x0)
	/go/src/github.com/openshift/azure-disk-csi-driver/vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_standard.go:161 +0x1916
sigs.k8s.io/cloud-provider-azure/pkg/provider.(*controllerCommon).DetachDisk(0xc000193450, 0xc0003786b7, 0x28, 0xc000378640, 0x9f, 0xc00060c570, 0x17, 0x27, 0xc0003fcce0)
	/go/src/github.com/openshift/azure-disk-csi-driver/vendor/sigs.k8s.io/cloud-provider-azure/pkg/provider/azure_controller_common.go:309 +0x1386
sigs.k8s.io/azuredisk-csi-driver/pkg/azuredisk.(*Driver).ControllerUnpublishVolume(0xc000120d80, 0x1c8ae58, 0xc000174690, 0xc00003c140, 0x0, 0x0, 0x0)
	/go/src/github.com/openshift/azure-disk-csi-driver/pkg/azuredisk/controllerserver.go:468 +0x385
github.com/container-storage-interface/spec/lib/go/csi._Controller_ControllerUnpublishVolume_Handler.func1(0x1c8ae58, 0xc000174690, 0x1959020, 0xc00003c140, 0x0, 0x0, 0x1a57801, 0x10)
	/go/src/github.com/openshift/azure-disk-csi-driver/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:5561 +0x89
sigs.k8s.io/azuredisk-csi-driver/pkg/csi-common.logGRPC(0x1c8ae58, 0xc000174690, 0x1959020, 0xc00003c140, 0xc0003fcbc0, 0xc0006a46a8, 0xc000758b80, 0x49cc86, 0x192e0c0, 0xc000174690)
	/go/src/github.com/openshift/azure-disk-csi-driver/pkg/csi-common/utils.go:98 +0x2ab
github.com/container-storage-interface/spec/lib/go/csi._Controller_ControllerUnpublishVolume_Handler(0x1a1c560, 0xc000120d80, 0x1c8ae58, 0xc000174690, 0xc0001abf20, 0x1b17268, 0x1c8ae58, 0xc000174690, 0xc000022b40, 0xbb)
	/go/src/github.com/openshift/azure-disk-csi-driver/vendor/github.com/container-storage-interface/spec/lib/go/csi/csi.pb.go:5563 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000602780, 0x1c9e1f8, 0xc00076f080, 0xc000118e00, 0xc000036270, 0x273b128, 0x0, 0x0, 0x0)
	/go/src/github.com/openshift/azure-disk-csi-driver/vendor/google.golang.org/grpc/server.go:1024 +0x522
google.golang.org/grpc.(*Server).handleStream(0xc000602780, 0x1c9e1f8, 0xc00076f080, 0xc000118e00, 0x0)
	/go/src/github.com/openshift/azure-disk-csi-driver/vendor/google.golang.org/grpc/server.go:1313 +0xd4c
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0001945b0, 0xc000602780, 0x1c9e1f8, 0xc00076f080, 0xc000118e00)
	/go/src/github.com/openshift/azure-disk-csi-driver/vendor/google.golang.org/grpc/server.go:722 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/openshift/azure-disk-csi-driver/vendor/google.golang.org/grpc/server.go:720 +0xa5

```

I could trace this back to some in-tree code that was added to fix the ASH build here: https://github.com/kubernetes/legacy-cloud-providers/commit/e7d1b0c6c90cba11798462ddfd1b2cf330218149

However I can make very little sense of this in my current azure stackhub setup. There is only a single disk and thus the extra slice logic does not work.

**What you expected to happen**:

I do not expect the driver to panic in that scenario and properly detach the disk.

**How to reproduce it**:

I'm running our [openshift origin e2e tests](https://github.com/openshift/origin) using:

> ./openshift-tests run openshift/csi --run="Dynamic.*multiVolume"

There will be several failures like that:

```
Failing tests:

External Storage [Driver: disk.csi.azure.com] [Testpattern: Dynamic PV (block volmode)] multiVolume [Slow] should access to two volumes with different volume mode and retain data across pod recreation on different node [LinuxOnly]
External Storage [Driver: disk.csi.azure.com] [Testpattern: Dynamic PV (block volmode)] multiVolume [Slow] should access to two volumes with the same volume mode and retain data across pod recreation on different node [LinuxOnly]
External Storage [Driver: disk.csi.azure.com] [Testpattern: Dynamic PV (block volmode)] multiVolume [Slow] should access to two volumes with the same volume mode and retain data across pod recreation on the same node [LinuxOnly]
External Storage [Driver: disk.csi.azure.com] [Testpattern: Dynamic PV (filesystem volmode)] multiVolume [Slow] should access to two volumes with the same volume mode and retain data across pod recreation on different node [LinuxOnly]
External Storage [Driver: disk.csi.azure.com] [Testpattern: Dynamic PV (filesystem volmode)] multiVolume [Slow] should access to two volumes with the same volume mode and retain data across pod recreation on the same node [LinuxOnly]
External Storage [Driver: disk.csi.azure.com] [Testpattern: Dynamic PV (filesystem volmode)] multiVolume [Slow] should concurrently access the single read-only volume from pods on the same node
External Storage [Driver: disk.csi.azure.com] [Testpattern: Dynamic PV (filesystem volmode)] multiVolume [Slow] should concurrently access the single volume from pods on the same node [LinuxOnly]
```

Running the tests triggers the above failure conditions.


**Anything else we need to know?**:

We got a ppe3 ASH deployment and we have deployed OpenShift over it, I can share the access if you want to take a closer look. 

**Environment**:
- Kubernetes version (use `kubectl version`): 

```
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.1-1380+120883fe7ff642-dirty", GitCommit:"120883fe7ff642c7de79ec1e533e4c63cf099eb0", GitTreeState:"dirty", BuildDate:"2021-06-12T07:23:02Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
```

- OS (e.g. from /etc/os-release):

```
NAME="Red Hat Enterprise Linux CoreOS"
VERSION="47.83.202104131913-0"
VERSION_ID="4.7"
OPENSHIFT_VERSION="4.7"
RHEL_VERSION="8.3"
PRETTY_NAME="Red Hat Enterprise Linux CoreOS 47.83.202104131913-0 (Ootpa)"
ID="rhcos"
ID_LIKE="rhel fedora"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8::coreos"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="OpenShift Container Platform"
REDHAT_BUGZILLA_PRODUCT_VERSION="4.7"
REDHAT_SUPPORT_PRODUCT="OpenShift Container Platform"
REDHAT_SUPPORT_PRODUCT_VERSION="4.7"
OSTREE_VERSION='47.83.202104131913-0'
```

- Kernel (e.g. `uname -a`): 4.18.0-240.8.1.el8_3.x86_64 #1 SMP Fri Dec 4 12:24:03 EST 2020 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: we installed the driver using our patched operator: https://github.com/openshift/azure-disk-csi-driver-operator/pull/26
- Others:

--- a/pkg/provider/azure_controller_common_test.go
+++ b/pkg/provider/azure_controller_common_test.go
@@ -68,7 +68,7 @@ func TestCommonAttachDisk(t *testing.T) {
 			nodeName:    "vm1",
 			diskName:    "disk-name",
 			existedDisk: nil,
-			expectedLun: 1,
+			expectedLun: 3,
 			expectedErr: false,
 		},
 		{
@@ -101,7 +101,7 @@ func TestCommonAttachDisk(t *testing.T) {
 					DiskState:  compute.Unattached,
 				},
 				Tags: testTags},
-			expectedLun: 1,
+			expectedLun: 3,
 			expectedErr: false,
 		},
 		{
@@ -372,7 +372,7 @@ func TestCommonUpdateVM(t *testing.T) {
 			desc:        "no error shall be returned if there's no matching disk according to given diskName",
 			vmList:      map[string]string{"vm1": "PowerState/Running"},
 			nodeName:    "vm1",
-			diskName:    "disk2",
+			diskName:    "diskx",
 			expectedErr: false,
 		},
 		{
@@ -428,7 +428,7 @@ func TestGetDiskLun(t *testing.T) {
 	}{
 		{
 			desc:        "LUN -1 and error shall be returned if diskName != disk.Name or diskURI != disk.Vhd.URI",
-			diskName:    "disk2",
+			diskName:    "diskx",
 			expectedLun: -1,
 			expectedErr: true,
 		},
@@ -480,7 +480,7 @@ func TestSetDiskLun(t *testing.T) {
 			nodeName:    "nodeName",
 			diskURI:     "diskURI",
 			diskMap:     map[string]*AttachDiskOptions{"diskURI": {}},
-			expectedLun: 1,
+			expectedLun: 3,
 			expectedErr: false,
 		},
 		{
@@ -544,9 +544,9 @@ func TestDisksAreAttached(t *testing.T) {
 		},
 		{
 			desc:             "proper attach map shall be returned if everything is good",
-			diskNames:        []string{"disk1", "disk2"},
+			diskNames:        []string{"disk1", "diskx"},
 			nodeName:         "vm1",
-			expectedAttached: map[string]bool{"disk1": true, "disk2": false},
+			expectedAttached: map[string]bool{"disk1": true, "diskx": false},
 			expectedErr:      false,
 		},
 	}

--- a/pkg/provider/azure_controller_standard.go
+++ b/pkg/provider/azure_controller_standard.go
@@ -158,11 +158,7 @@ func (as *availabilitySet) DetachDisk(nodeName types.NodeName, diskMap map[strin
 				(disk.ManagedDisk != nil && diskURI != "" && strings.EqualFold(*disk.ManagedDisk.ID, diskURI)) {
 				// found the disk
 				klog.V(2).Infof("azureDisk - detach disk: name %q uri %q", diskName, diskURI)
-				if strings.EqualFold(as.cloud.Environment.Name, consts.AzureStackCloudName) && !as.Config.DisableAzureStackCloud {
-					disks = append(disks[:i], disks[i+1:]...)
-				} else {
-					disks[i].ToBeDetached = to.BoolPtr(true)
-				}
+				disks[i].ToBeDetached = to.BoolPtr(true)
 				bFoundDisk = true
 			}
 		}
@@ -171,6 +167,17 @@ func (as *availabilitySet) DetachDisk(nodeName types.NodeName, diskMap map[strin
 	if !bFoundDisk {
 		// only log here, next action is to update VM status with original meta data
 		klog.Errorf("detach azure disk on node(%s): disk list(%s) not found", nodeName, diskMap)
+	} else {
+		if strings.EqualFold(as.cloud.Environment.Name, consts.AzureStackCloudName) && !as.Config.DisableAzureStackCloud {
+			// Azure stack does not support ToBeDetached flag, use original way to detach disk
+			newDisks := []compute.DataDisk{}
+			for _, disk := range disks {
+				if !to.Bool(disk.ToBeDetached) {
+					newDisks = append(newDisks, disk)
+				}
+			}
+			disks = newDisks
+		}
 	}
 
 	newVM := compute.VirtualMachineUpdate{

--- a/pkg/provider/azure_controller_standard_test.go
+++ b/pkg/provider/azure_controller_standard_test.go
@@ -17,6 +17,7 @@ limitations under the License.
 package provider
 
 import (
+	"fmt"
 	"net/http"
 	"testing"
 	"time"
@@ -126,7 +127,7 @@ func TestStandardDetachDisk(t *testing.T) {
 	testCases := []struct {
 		desc          string
 		nodeName      types.NodeName
-		diskName      string
+		disks         []string
 		isDetachFail  bool
 		expectedError bool
 	}{
@@ -138,13 +139,19 @@ func TestStandardDetachDisk(t *testing.T) {
 		{
 			desc:          "no error shall be returned if there's no corresponding disk",
 			nodeName:      "vm1",
-			diskName:      "disk2",
+			disks:         []string{"diskx"},
 			expectedError: false,
 		},
 		{
 			desc:          "no error shall be returned if there's a corresponding disk",
 			nodeName:      "vm1",
-			diskName:      "disk1",
+			disks:         []string{"disk1"},
+			expectedError: false,
+		},
+		{
+			desc:          "no error shall be returned if there's 2 corresponding disks",
+			nodeName:      "vm1",
+			disks:         []string{"disk1", "disk2"},
 			expectedError: false,
 		},
 		{
@@ -167,17 +174,20 @@ func TestStandardDetachDisk(t *testing.T) {
 		if test.isDetachFail {
 			mockVMsClient.EXPECT().Update(gomock.Any(), testCloud.ResourceGroup, gomock.Any(), gomock.Any(), gomock.Any()).Return(&retry.Error{HTTPStatusCode: http.StatusNotFound, RawError: cloudprovider.InstanceNotFound}).AnyTimes()
 		} else {
-			mockVMsClient.EXPECT().Update(gomock.Any(), testCloud.ResourceGroup, gomock.Any(), gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+			mockVMsClient.EXPECT().Update(gomock.Any(), testCloud.ResourceGroup, "vm1", gomock.Any(), "detach_disk").Return(nil).AnyTimes()
 		}
 
-		diskMap := map[string]string{
-			"uri": test.diskName,
+		diskMap := map[string]string{}
+		for _, diskName := range test.disks {
+			diskURI := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/disks/%s",
+				testCloud.SubscriptionID, testCloud.ResourceGroup, diskName)
+			diskMap[diskURI] = diskName
 		}
 		err := vmSet.DetachDisk(test.nodeName, diskMap)
 		assert.Equal(t, test.expectedError, err != nil, "TestCase[%d]: %s", i, test.desc)
-		if !test.expectedError && test.diskName != "" {
+		if !test.expectedError && len(test.disks) > 0 {
 			dataDisks, _, err := vmSet.GetDataDisks(test.nodeName, azcache.CacheReadTypeDefault)
-			assert.Equal(t, true, len(dataDisks) == 1, "TestCase[%d]: %s, err: %v", i, test.desc, err)
+			assert.Equal(t, true, len(dataDisks) == 3, "TestCase[%d]: %s, err: %v", i, test.desc, err)
 		}
 	}
 }
@@ -237,7 +247,7 @@ func TestStandardUpdateVM(t *testing.T) {
 		assert.Equal(t, test.expectedError, err != nil, "TestCase[%d]: %s", i, test.desc)
 		if !test.expectedError && test.diskName != "" {
 			dataDisks, _, err := vmSet.GetDataDisks(test.nodeName, azcache.CacheReadTypeDefault)
-			assert.Equal(t, true, len(dataDisks) == 1, "TestCase[%d]: %s, err: %v", i, test.desc, err)
+			assert.Equal(t, true, len(dataDisks) == 3, "TestCase[%d]: %s, err: %v", i, test.desc, err)
 		}
 	}
 }

--- a/pkg/provider/azure_controller_standard_test.go
+++ b/pkg/provider/azure_controller_standard_test.go
@@ -279,6 +279,14 @@ func TestGetDataDisks(t *testing.T) {
 					Lun:  to.Int32Ptr(0),
 					Name: to.StringPtr("disk1"),
 				},
+				{
+					Lun:  to.Int32Ptr(1),
+					Name: to.StringPtr("disk2"),
+				},
+				{
+					Lun:  to.Int32Ptr(2),
+					Name: to.StringPtr("disk3"),
+				},
 			},
 			expectedError: false,
 			crt:           azcache.CacheReadTypeDefault,
@@ -290,6 +298,14 @@ func TestGetDataDisks(t *testing.T) {
 				{
 					Lun:  to.Int32Ptr(0),
 					Name: to.StringPtr("disk1"),
+				},
+				{
+					Lun:  to.Int32Ptr(1),
+					Name: to.StringPtr("disk2"),
+				},
+				{
+					Lun:  to.Int32Ptr(2),
+					Name: to.StringPtr("disk3"),
 				},
 			},
 			expectedError: false,

--- a/pkg/provider/azure_controller_vmss.go
+++ b/pkg/provider/azure_controller_vmss.go
@@ -162,11 +162,7 @@ func (ss *ScaleSet) DetachDisk(nodeName types.NodeName, diskMap map[string]strin
 				(disk.ManagedDisk != nil && diskURI != "" && strings.EqualFold(*disk.ManagedDisk.ID, diskURI)) {
 				// found the disk
 				klog.V(2).Infof("azureDisk - detach disk: name %q uri %q", diskName, diskURI)
-				if strings.EqualFold(ss.cloud.Environment.Name, consts.AzureStackCloudName) && !ss.Config.DisableAzureStackCloud {
-					disks = append(disks[:i], disks[i+1:]...)
-				} else {
-					disks[i].ToBeDetached = to.BoolPtr(true)
-				}
+				disks[i].ToBeDetached = to.BoolPtr(true)
 				bFoundDisk = true
 			}
 		}
@@ -175,6 +171,17 @@ func (ss *ScaleSet) DetachDisk(nodeName types.NodeName, diskMap map[string]strin
 	if !bFoundDisk {
 		// only log here, next action is to update VM status with original meta data
 		klog.Errorf("detach azure disk on node(%s): disk list(%s) not found", nodeName, diskMap)
+	} else {
+		if strings.EqualFold(ss.cloud.Environment.Name, consts.AzureStackCloudName) && !ss.Config.DisableAzureStackCloud {
+			// Azure stack does not support ToBeDetached flag, use original way to detach disk
+			newDisks := []compute.DataDisk{}
+			for _, disk := range disks {
+				if !to.Bool(disk.ToBeDetached) {
+					newDisks = append(newDisks, disk)
+				}
+			}
+			disks = newDisks
+		}
 	}
 
 	newVM := compute.VirtualMachineScaleSetVM{

--- a/pkg/provider/azure_controller_vmss_test.go
+++ b/pkg/provider/azure_controller_vmss_test.go
@@ -383,7 +383,7 @@ func TestUpdateVMWithVMSS(t *testing.T) {
 
 		if !test.expectedErr {
 			dataDisks, _, err := ss.GetDataDisks(test.vmssvmName, azcache.CacheReadTypeDefault)
-			assert.Equal(t, true, len(dataDisks) == 3, "TestCase[%d]: %s, actual data disk num: %d, err: %v", i, test.desc, len(dataDisks), err)
+			assert.Equal(t, true, len(dataDisks) == 1, "TestCase[%d]: %s, actual data disk num: %d, err: %v", i, test.desc, len(dataDisks), err)
 		}
 	}
 }

--- a/pkg/provider/azure_controller_vmss_test.go
+++ b/pkg/provider/azure_controller_vmss_test.go
@@ -44,7 +44,7 @@ func TestAttachDiskWithVMSS(t *testing.T) {
 		desc           string
 		vmssName       types.NodeName
 		vmssvmName     types.NodeName
-		existedDisk    compute.Disk
+		disks          []string
 		vmList         map[string]string
 		vmssVMList     []string
 		isManagedDisk  bool
@@ -57,17 +57,26 @@ func TestAttachDiskWithVMSS(t *testing.T) {
 			vmssName:       "vm1",
 			vmssvmName:     "vm1",
 			isManagedDisk:  false,
-			existedDisk:    compute.Disk{Name: to.StringPtr("disk-name")},
+			disks:          []string{"disk-name"},
 			expectedErr:    true,
 			expectedErrMsg: fmt.Errorf("not a vmss instance"),
 		},
 		{
-			desc:          "no error shall be returned if everything is good with managed disk",
+			desc:          "no error shall be returned if everything is good with one managed disk",
 			vmssVMList:    []string{"vmss00-vm-000000", "vmss00-vm-000001", "vmss00-vm-000002"},
 			vmssName:      "vmss00",
 			vmssvmName:    "vmss00-vm-000000",
 			isManagedDisk: true,
-			existedDisk:   compute.Disk{Name: to.StringPtr("disk-name")},
+			disks:         []string{"disk-name"},
+			expectedErr:   false,
+		},
+		{
+			desc:          "no error shall be returned if everything is good with 2 managed disks",
+			vmssVMList:    []string{"vmss00-vm-000000", "vmss00-vm-000001", "vmss00-vm-000002"},
+			vmssName:      "vmss00",
+			vmssvmName:    "vmss00-vm-000000",
+			isManagedDisk: true,
+			disks:         []string{"disk-name", "disk-name2"},
 			expectedErr:   false,
 		},
 		{
@@ -76,7 +85,7 @@ func TestAttachDiskWithVMSS(t *testing.T) {
 			vmssName:      "vmss00",
 			vmssvmName:    "vmss00-vm-000000",
 			isManagedDisk: false,
-			existedDisk:   compute.Disk{Name: to.StringPtr("disk-name")},
+			disks:         []string{"disk-name"},
 			expectedErr:   false,
 		},
 		{
@@ -85,7 +94,7 @@ func TestAttachDiskWithVMSS(t *testing.T) {
 			vmssName:       fakeStatusNotFoundVMSSName,
 			vmssvmName:     "vmss00-vm-000000",
 			isManagedDisk:  false,
-			existedDisk:    compute.Disk{Name: to.StringPtr("disk-name")},
+			disks:          []string{"disk-name"},
 			expectedErr:    true,
 			expectedErrMsg: fmt.Errorf("Retriable: false, RetryAfter: 0s, HTTPStatusCode: 404, RawError: %w", fmt.Errorf("instance not found")),
 		},
@@ -126,19 +135,20 @@ func TestAttachDiskWithVMSS(t *testing.T) {
 			mockVMSSVMClient.EXPECT().Update(gomock.Any(), testCloud.ResourceGroup, scaleSetName, gomock.Any(), gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
 		}
 
-		diskURI := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/disks/%s",
-			testCloud.SubscriptionID, testCloud.ResourceGroup, *test.existedDisk.Name)
+		diskMap := map[string]*AttachDiskOptions{}
+		for i, diskName := range test.disks {
+			options := AttachDiskOptions{
+				lun:                     int32(i),
+				isManagedDisk:           test.isManagedDisk,
+				diskName:                diskName,
+				cachingMode:             compute.CachingTypesReadWrite,
+				diskEncryptionSetID:     "",
+				writeAcceleratorEnabled: true,
+			}
 
-		options := AttachDiskOptions{
-			lun:                     0,
-			isManagedDisk:           test.isManagedDisk,
-			diskName:                "disk-name",
-			cachingMode:             compute.CachingTypesReadWrite,
-			diskEncryptionSetID:     "",
-			writeAcceleratorEnabled: true,
-		}
-		diskMap := map[string]*AttachDiskOptions{
-			diskURI: &options,
+			diskURI := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/disks/%s",
+				testCloud.SubscriptionID, testCloud.ResourceGroup, diskName)
+			diskMap[diskURI] = &options
 		}
 		err = ss.AttachDisk(test.vmssvmName, diskMap)
 		assert.Equal(t, test.expectedErr, err != nil, "TestCase[%d]: %s, return error: %v", i, test.desc, err)
@@ -160,7 +170,7 @@ func TestDetachDiskWithVMSS(t *testing.T) {
 		vmssVMList     []string
 		vmssName       types.NodeName
 		vmssvmName     types.NodeName
-		existedDisk    compute.Disk
+		disks          []string
 		expectedErr    bool
 		expectedErrMsg error
 	}{
@@ -169,7 +179,7 @@ func TestDetachDiskWithVMSS(t *testing.T) {
 			vmssVMList:     []string{"vmss-vm-000001"},
 			vmssName:       "vm1",
 			vmssvmName:     "vm1",
-			existedDisk:    compute.Disk{Name: to.StringPtr(diskName)},
+			disks:          []string{diskName},
 			expectedErr:    true,
 			expectedErrMsg: fmt.Errorf("not a vmss instance"),
 		},
@@ -178,7 +188,15 @@ func TestDetachDiskWithVMSS(t *testing.T) {
 			vmssVMList:  []string{"vmss00-vm-000000", "vmss00-vm-000001", "vmss00-vm-000002"},
 			vmssName:    "vmss00",
 			vmssvmName:  "vmss00-vm-000000",
-			existedDisk: compute.Disk{Name: to.StringPtr(diskName)},
+			disks:       []string{diskName},
+			expectedErr: false,
+		},
+		{
+			desc:        "no error shall be returned if everything is good",
+			vmssVMList:  []string{"vmss00-vm-000000", "vmss00-vm-000001", "vmss00-vm-000002"},
+			vmssName:    "vmss00",
+			vmssvmName:  "vmss00-vm-000000",
+			disks:       []string{diskName, "disk2"},
 			expectedErr: false,
 		},
 		{
@@ -186,7 +204,7 @@ func TestDetachDiskWithVMSS(t *testing.T) {
 			vmssVMList:     []string{"vmss00-vm-000000", "vmss00-vm-000001", "vmss00-vm-000002"},
 			vmssName:       fakeStatusNotFoundVMSSName,
 			vmssvmName:     "vmss00-vm-000000",
-			existedDisk:    compute.Disk{Name: to.StringPtr(diskName)},
+			disks:          []string{diskName},
 			expectedErr:    true,
 			expectedErrMsg: fmt.Errorf("Retriable: false, RetryAfter: 0s, HTTPStatusCode: 404, RawError: %w", fmt.Errorf("instance not found")),
 		},
@@ -195,7 +213,7 @@ func TestDetachDiskWithVMSS(t *testing.T) {
 			vmssVMList:  []string{"vmss00-vm-000000", "vmss00-vm-000001", "vmss00-vm-000002"},
 			vmssName:    "vmss00",
 			vmssvmName:  "vmss00-vm-000000",
-			existedDisk: compute.Disk{Name: to.StringPtr("disk-name-err")},
+			disks:       []string{"disk-name-err"},
 			expectedErr: false,
 		},
 	}
@@ -224,10 +242,20 @@ func TestDetachDiskWithVMSS(t *testing.T) {
 						},
 					},
 				},
-				DataDisks: &[]compute.DataDisk{{
-					Lun:  to.Int32Ptr(0),
-					Name: to.StringPtr(diskName),
-				}},
+				DataDisks: &[]compute.DataDisk{
+					{
+						Lun:  to.Int32Ptr(0),
+						Name: to.StringPtr(diskName),
+					},
+					{
+						Lun:  to.Int32Ptr(1),
+						Name: to.StringPtr("disk2"),
+					},
+					{
+						Lun:  to.Int32Ptr(2),
+						Name: to.StringPtr("disk3"),
+					},
+				},
 			}
 		}
 		mockVMSSVMClient := testCloud.VirtualMachineScaleSetVMsClient.(*mockvmssvmclient.MockInterface)
@@ -238,8 +266,11 @@ func TestDetachDiskWithVMSS(t *testing.T) {
 			mockVMSSVMClient.EXPECT().Update(gomock.Any(), testCloud.ResourceGroup, scaleSetName, gomock.Any(), gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
 		}
 
-		diskMap := map[string]string{
-			diskName: *test.existedDisk.Name,
+		diskMap := map[string]string{}
+		for _, diskName := range test.disks {
+			diskURI := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/disks/%s",
+				testCloud.SubscriptionID, testCloud.ResourceGroup, diskName)
+			diskMap[diskURI] = diskName
 		}
 		err = ss.DetachDisk(test.vmssvmName, diskMap)
 		assert.Equal(t, test.expectedErr, err != nil, "TestCase[%d]: %s, err: %v", i, test.desc, err)
@@ -249,7 +280,7 @@ func TestDetachDiskWithVMSS(t *testing.T) {
 
 		if !test.expectedErr {
 			dataDisks, _, err := ss.GetDataDisks(test.vmssvmName, azcache.CacheReadTypeDefault)
-			assert.Equal(t, true, len(dataDisks) == 1, "TestCase[%d]: %s, actual data disk num: %d, err: %v", i, test.desc, len(dataDisks), err)
+			assert.Equal(t, true, len(dataDisks) == 3, "TestCase[%d]: %s, actual data disk num: %d, err: %v", i, test.desc, len(dataDisks), err)
 		}
 	}
 }
@@ -352,7 +383,7 @@ func TestUpdateVMWithVMSS(t *testing.T) {
 
 		if !test.expectedErr {
 			dataDisks, _, err := ss.GetDataDisks(test.vmssvmName, azcache.CacheReadTypeDefault)
-			assert.Equal(t, true, len(dataDisks) == 1, "TestCase[%d]: %s, actual data disk num: %d, err: %v", i, test.desc, len(dataDisks), err)
+			assert.Equal(t, true, len(dataDisks) == 3, "TestCase[%d]: %s, actual data disk num: %d, err: %v", i, test.desc, len(dataDisks), err)
 		}
 	}
 }

--- a/pkg/provider/azure_instances_test.go
+++ b/pkg/provider/azure_instances_test.go
@@ -76,10 +76,20 @@ func setTestVirtualMachines(c *Cloud, vmList map[string]string, isDataDisksFull 
 			},
 		}
 		if !isDataDisksFull {
-			vm.StorageProfile.DataDisks = &[]compute.DataDisk{{
-				Lun:  to.Int32Ptr(0),
-				Name: to.StringPtr("disk1"),
-			}}
+			vm.StorageProfile.DataDisks = &[]compute.DataDisk{
+				{
+					Lun:  to.Int32Ptr(0),
+					Name: to.StringPtr("disk1"),
+				},
+				{
+					Lun:  to.Int32Ptr(1),
+					Name: to.StringPtr("disk2"),
+				},
+				{
+					Lun:  to.Int32Ptr(2),
+					Name: to.StringPtr("disk3"),
+				},
+			}
 		} else {
 			dataDisks := make([]compute.DataDisk, maxLUN)
 			for i := 0; i < maxLUN; i++ {

